File: Graph Advanced/GraphSearchPractice.py
# Practice 10. Graph Representation 
import sys
from collections import deque
CONSTRUCT = 'C'
IS_ADJACENT = 'I'
GET_NEIGHBORS = 'N'
BFS = 'B'
DFS = 'D'
REACHABILITY = 'R'
TOPOLOGICAL_SORT = 'T'
SHORTEST_PATH = 'S'

class Graph:
  # TODO. Define a constructor and proper methods
  def __init__(self):
    self.adjacency = {}

  def AddVertice(self, v1, v2, w):
    if self.adjacency.get(v1) == None:
      self.adjacency[v1] = []
    if self.adjacency.get(v2) == None:
      self.adjacency[v2] = []
    self.adjacency[v1].append((v2, w))

  def IsAdjacent(self, v1, v2):
    if v2 in map(lambda x: x[0], self.adjacency[v1]):
      return 'T'
    else:
      return 'F'

  def Neighbors(self, v):
    adjacency = []
    for tmp in self.adjacency[v]:
      adjacency.append(str(tmp[0]))
    return adjacency
  
  def DepthFirstSearch(self, st):
    visited = []
    s = deque()
    s.append(st)
    while s:
      u = s.pop()
      if u not in visited:
        visited.append(u)
        for v in self.adjacency[u]:
          if v[0] not in visited:
            s.append(v[0])
    return " ".join(list(map(str, visited)))

  def BreadthFirstSearch(self, st):
    visited = []
    q = deque()
    visited.append(st)
    q.append(st)
    while q:
      u = q.popleft()
      for v in self.adjacency[u]:
        if v[0] not in visited:
          visited.append(v[0])
          q.append(v[0])
    return " ".join(list(map(str, visited)))

  def PrintGraph(self):
    for i in sorted(self.adjacency.keys()):
      string = ""
      for j in self.adjacency[i]:
        string += f"{j} -> "
      string += "|"
      print(f"{i} -> {string}")

  def InDegree(self, u):
    cnt = 0
    for i in self.adjacency.keys():
      lst = [j[0] for j in self.adjacency[i]]
      cnt += lst.count(u)
    return cnt

  def TopologicalSort(self):
    indegree = {i : self.InDegree(i) for i in self.adjacency.keys()}
    t = []
    q = deque()
    for v in self.adjacency.keys():
      if indegree[v] == 0:
        q.append(v)
    while q:
      u = q.popleft()
      t.append(u)
      lst = [i[0] for i in self.adjacency[u]]
      for v in lst:
        indegree[v] -= 1
        if indegree[v] == 0:
          q.append(v)
    return t

if __name__ == "__main__":  
  g = Graph()
  with open('input.txt', 'r') as inFile:
    lines = inFile.readlines()
  with open('output.txt', 'w') as outFile:
    i = 0
    while i < len(lines):
      words = lines[i].split()
      op = words[0]

      if op == CONSTRUCT:
        if len(words) != 3:
          raise Exception("CONSTRUCT: invalid input")
        
        n, m = int(words[1]), int(words[2])
        for _ in range(m):
          i += 1
          v1, v2, w = map(int, lines[i].split())
          g.AddVertice(v1, v2, w)

      elif op == IS_ADJACENT:
        if len(words) != 3:
          raise Exception("IS_ADJACENT: invalid input")
        u, v = int(words[1]), int(words[2])
        outFile.write(f'{u} {v} {g.IsAdjacent(u, v)}\n')

      elif op == GET_NEIGHBORS:
        if len(words) != 2:
          raise Exception("GET_NEIGHBORS: invalid input")
        u = int(words[1])
        outFile.write(" ".join(g.Neighbors(u)))
        outFile.write('\n')

      elif op == DFS:
        outFile.write(g.DepthFirstSearch(int(words[1])))
        outFile.write("\n")

      elif op == BFS:
        outFile.write(g.BreadthFirstSearch(int(words[1])))
        outFile.write("\n")

      elif op == TOPOLOGICAL_SORT:
        outFile.write(" ".join(list(map(str, g.TopologicalSort()))))
        outFile.write("\n")
      
      elif op == SHORTEST_PATH:
        pass

      else:
        raise Exception("Undefined operator")
      
      i += 1
  g.PrintGraph()

File: Graph Advanced/GraphSearchPractice.cpp
// Practice 10. Graph Representation
#include <algorithm>
#include <cassert>
#include <climits>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <queue> // BFS
#include <sstream>
#include <stack> // DFS
#include <tuple>
#include <vector>
#include <map>
using namespace std;
const char CONSTRUCT = 'C';
const char IS_ADJACENT = 'I';
const char GET_NEIGHBORS = 'N';
const char BFS = 'B';
const char DFS = 'D';
const char REACHABLITY = 'R';
const char TOPOLOGICAL_SORT = 'T';
const char SHORTEST_PATH = 'S';

class Graph {
public:
int nodes;
map<int, vector<pair<int, int> > > adjacency;
// TODO. Define a constructor, a destructor, and proper methods

void AddVertice(int, int, int);
string IsAdjacent(int, int);
string Neighbors(int);
string BreadthFirstSearch(int);
string DepthFirstSearch(int);
string TopologicalSort();
int InDegree(int);
void PrintGraph();
};

void Graph::AddVertice(int v1, int v2, int weight) {
  adjacency[v1];
  adjacency[v2];
  adjacency[v1].push_back(make_pair(v2, weight));
}

string Graph::IsAdjacent(int v1, int v2) {
  vector<pair<int, int> > from = adjacency[v1];
  for (vector<pair<int, int> >::iterator iter=from.begin(); iter!=from.end(); iter++) {
    if (iter->first == v2)
      return "T";
  }
  return "F";
}

string Graph::Neighbors(int v) {
  string neighbors = "";
  for (int i=0; i<adjacency[v].size(); i++)
    neighbors += to_string(adjacency[v][i].first) + " ";
  return neighbors;
}

string Graph::DepthFirstSearch(int st) {
  vector<int> visited;
  stack<int> s;
  string r = "";
  s.push(st);
  while (!s.empty()) {
    int u = s.top();
    s.pop();
    if (find(visited.begin(), visited.end(), u) == visited.end()) {
      visited.push_back(u);
      for (int i=0; i<adjacency[u].size(); i++) {
        if (find(visited.begin(), visited.end(), adjacency[u][i].first) == visited.end())
          s.push(adjacency[u][i].first);
      }
    }
  }

  for (int i=0; i<visited.size(); i++)
    r += to_string(visited[i]) + " ";
  return r;
}

string Graph::BreadthFirstSearch(int st) {
  vector<int> visited;
  queue<int> q;
  string r = "";
  visited.push_back(st);
  q.push(st);
  while (!q.empty()) {
    int u = q.front();
    q.pop();
    for (int i=0; i<adjacency[u].size(); i++) {
      if (find(visited.begin(), visited.end(), adjacency[u][i].first) == visited.end()) {
        visited.push_back(adjacency[u][i].first);
        q.push(adjacency[u][i].first);
      }
    }
  }
  for (int i=0; i<visited.size(); i++)
    r += to_string(visited[i]) + " ";
  return r;
}

int Graph::InDegree(int u) {
  int cnt = 0;
  vector<int> temp;
  for (map<int, vector<pair<int, int> > >::iterator iter=adjacency.begin(); iter!=adjacency.end(); iter++) {
    for (int i=0; i<adjacency[iter->first].size(); i++) {
      temp.push_back(adjacency[iter->first][i].first);
    }
  }
  cnt += count(temp.begin(), temp.end(), u);
  return cnt;
}

string Graph::TopologicalSort() {
  map<int, int> indegree;
  for (map<int, vector<pair<int, int> > >::iterator iter=adjacency.begin(); iter!=adjacency.end(); iter++) {
    indegree[iter->first] = Graph::InDegree(iter->first);
  }
  vector<int> t;
  queue<int> q;
  for (map<int, vector<pair<int, int> > >::iterator iter=adjacency.begin(); iter!=adjacency.end(); iter++) {
    if (indegree[iter->first] == 0)
      q.push(iter->first);
  }
  while (!q.empty()) {
    int u = q.front();
    q.pop();
    t.push_back(u);
    for (vector<pair<int, int> >::iterator iter=adjacency[u].begin(); iter!=adjacency[u].end(); iter++) {
      indegree[iter->first]--;
      if (indegree[iter->first] == 0)
        q.push(iter->first);
    }
  }
  string r = "";
  for (int i=0; i<t.size(); i++)
    r += to_string(t[i]) + " ";
  return r;
}

void Graph::PrintGraph() {
  for (map<int, vector<pair<int, int> > >::iterator iter=adjacency.begin(); iter!=adjacency.end(); iter++) {
    string r = to_string(iter->first) + " -> ";
    for (vector<pair<int, int> >::iterator it=iter->second.begin(); it!=iter->second.end(); it++)
      r += "(" + to_string(it->first) + ", " + to_string(it->second) + ") -> ";
    r += "|";
    cout << r << endl;
  }
}

int main() {
  Graph g;
  ifstream inFile("input.txt");
  ofstream outFile("output.txt");
  string line;
  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(1));
    int u, v, wgt, nVertices, nEdges, cnt;
    vector<int> p, nbrs;
    switch(op)
    {
      case CONSTRUCT:
        if (!(iss >> nVertices >> nEdges)) {
          cerr<<"CONSTRUCT: invalid input"<<endl;
          exit(1);
        }
        // TODO. Construct a graph
        for (int i=0; i<nEdges; i++) {
          getline(inFile, line);
          istringstream is_(line);
          is_ >> u >> v >> wgt;
          g.AddVertice(u, v, wgt);
        }
        break;

      case IS_ADJACENT:
        if (!(iss >> u >> v)) {
          cerr<<"isAdjacent: invalid input"<<endl;
          exit(1);
        }
        // TODO. Check if edge (u, v) exists in the graph
        outFile << u << " " << v << " " << g.IsAdjacent(u, v) << endl;
        break;

      case GET_NEIGHBORS:
        if (!(iss >> u)) {
          cerr<<"getNeighbors: invalid input"<<endl;
          exit(1);
        }
        // TODO. Get all the neighbors of u
        outFile << g.Neighbors(u) << endl;
        break;

      case DFS:
        if (!(iss >> u)) {
          cerr<<"DFS: invalid input"<<endl;
          exit(1);
        }
        outFile << g.DepthFirstSearch(u) << endl;
        break;

      case BFS:
        if (!(iss >> u)) {
          cerr<<"BFS: invalid input"<<endl;
          exit(1);
        }
        outFile << g.BreadthFirstSearch(u) << endl;
        break;
      
      case TOPOLOGICAL_SORT:
        outFile << g.TopologicalSort() << endl;
        break;
      
      default:
        cerr<<"Undefined operator"<<endl;
        exit(1);
    }
  }
  g.PrintGraph();
  outFile.close();
  inFile.close();
}
